nav2_costmap_2d/include/nav2_costmap_2d/image_processing.hpp

```diff
@@ -0,0 +1,880 @@
+// Copyright (c) 2021 Andrey Ryzhikov
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_COSTMAP_2D__IMAGE_PROCESSING_HPP_
+#define NAV2_COSTMAP_2D__IMAGE_PROCESSING_HPP_
+
+#include "image.hpp"
+#include <algorithm>
+
+namespace nav2_costmap_2d
+{
+
+/// Pixels connectivity type (is the way in which pixels in image relate to
+/// their neighbors)
+enum class ConnectivityType: int
+{
+  /// neighbors pixels are connected horizontally and vertically
+  Way4 = 4,
+  /// neighbors pixels are connected horizontally, vertically and diagonally
+  Way8 = 8
+};
+
+/**
+ * @brief Calculate truncated histogram of image.
+ *
+ * Creates a histogram of image_max bins.
+ * Bin with index i keep min of (<count of the number of pixels with value i>, bin_max).
+ * This truncation avoids overflow and is acceptable in the problem being solved.
+ * For example, the image with pixel type uint16_t may have 100'000 pixels equal to 0
+ * (100'000 > std::numeric_limits<uint16_t>). In this case, an overflow will occur when
+ * calculating a traditional histogram with bins of the uint16_t type. But in this function,
+ * the bin value will increase to the bin_max value, then stop. Overflow will not happen.
+ * @tparam T image pixel type
+ * @param image source image
+ * @param image_max max image pixel value
+ * @param bin_max max histogram bin value
+ * @return vector of histogram bins
+ * @warning If source contains a pixel with a value, large then image_max,
+ * the behavior is undefined
+*/
+template<class T, class Bin>
+std::vector<Bin> histogram(const Image<T> & image, T image_max, Bin bin_max);
```

Comment:
No need to declare right above here, the implementation should be enough. Mark as static or inline to have the floating function (probably static, unless you have multiple threads accessing it, this looks a bit large for inline).

Reply:
Ok, I will delete the declaration. But it's not clear to me why add `static` or `inline` in this case?
```cpp
template<class T, class Bin>
inline std::vector<Bin> histogram(const Image<T> & image, T image_max, Bin bin_max)
{
...
}
```
`inline` doesn't seem to have any effect here. The `static` modifier will not be ignored by the compiler, unlike `inline`. But I don't understand why `static` is needed in this case. Can you explain?

Reply:
I'd have to clone and play around, but usually if you have  floating function, it needs some modifier to link correctly if used in another file. I think if you tried without marking as `static`, you'd end up with linking errors. If that is not the case, then no modifier is required.